x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -34,19 +37,33 @@
     private static final int COPY_BUFFER_SIZE = 8192;
 
     private final CacheService cacheService;
+    private final SnapshotId snapshotId;
+    private final IndexId indexId;
+    private final ShardId shardId;
     private final Path cacheDir;
 
-    public CacheDirectory(Directory in, CacheService cacheService, Path cacheDir) throws IOException {
+    public CacheDirectory(Directory in, CacheService cacheService, Path cacheDir, SnapshotId snapshotId, IndexId indexId, ShardId shardId)
+        throws IOException {
         super(in);
         this.cacheService = Objects.requireNonNull(cacheService);
         this.cacheDir = Files.createDirectories(cacheDir);
+        this.snapshotId = Objects.requireNonNull(snapshotId);
+        this.indexId = Objects.requireNonNull(indexId);
+        this.shardId = Objects.requireNonNull(shardId);
+    }
+
+    private CacheKey createCacheKey(String fileName) {
+        return new CacheKey(snapshotId, indexId, shardId, fileName);
     }
 
     public void close() throws IOException {
         super.close();
         // Ideally we could let the cache evict/remove cached files by itself after the
         // directory has been closed.
-        cacheService.removeFromCache(key -> key.startsWith(cacheDir.toString()));
+        cacheService.removeFromCache(key ->
+            Objects.equals(key.getSnapshotId(), snapshotId) &&
```

Comment:
Maybe move this lambda to a method `CacheKey#belongsTo(SnapshotId, IndexId, ShardId)` (not sure about the name but that's the best I've got)

Reply:
Good idea. I kept the name and added a test in e8afbe7956be8027f056d70372920e5db11ca7ef